[HOPSWORKS-2996] Add Git integration documentation (#1036)
logicalclocks/hopsworks: docs/compute/git.md

# Git integration 

> The Git integration allows users to manage and version control their feature engineering codebase and workflows using Git directly from the Hopsworks UI. The integration currently supports repositories hosted on Github, Gitlab and Bitbucket. 

Reviewer: ```suggestion
The Git integration allows users to manage and version control their codebase and workflows using Git directly from the Hopsworks UI. The integration currently supports repositories hosted on Github, Gitlab and Bitbucket. 
```

Reviewer: True for now, but the focus is shifting towards an e2e platform so the codebase is not strictly feature engineering.

!!! warning "Beta"
    The feature is currently in Beta and will be improved in the upcoming releases.

## Credentials

> When you perform Git operations on Hopsworks that need to interact with the remote repository, Hopsworks relies on the Git HTTPs protocol to perform those operations. Authentication with the remote repository happens through a token generated by the Git repository hosting service (Github, Gitlab, Bitbucket).

Reviewer: ```suggestion
When you perform Git operations on Hopsworks that need to interact with the remote repository, Hopsworks relies on the Git HTTPS protocol to perform those operations. Authentication with the remote repository happens through a token generated by the Git repository hosting service (Github, Gitlab, Bitbucket).
```

Documentation on how to generate a token for the supported Git hosting services is available here:

- [Github](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
- [Gitlab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html)
- [Bitbucket](https://confluence.atlassian.com/bitbucketserver/http-access-tokens-939515499.html)

> Once the token has been generate, you need to provided it to Hopsworks so that it can be used to authenticate Git operations. In the _Account Settings_ page you can find the _Git Providers_ section. The Git provider section displays which providers have been already configured and can be used to clone new repositories.

Reviewer: ```suggestion
Once the token has been generated, you need to provide it to Hopsworks so that it can be used to authenticate Git operations. In the _Account Settings_ page you can find the _Git Providers_ section. The Git provider section displays which providers have been already configured and can be used to clone new repositories.
```

<p align="center">
  <figure>
    <a  href="../../assets/images/git/git_provider_1.png">
      <img src="../../assets/images/git/git_provider_1.png" alt="Git provider configuration list">
    </a>
    <figcaption>Git provider configuration list</figcaption>
  </figure>
</p>

You can click on the _Edit Configuration_ to change a provider username or token, or to configure a new provider.

<p align="center">
  <figure>
    <a  href="../../assets/images/git/git_provider_2.png">
      <img src="../../assets/images/git/git_provider_2.png" alt="Git provider configuration">
    </a>
    <figcaption>Git provider configuration</figcaption>
  </figure>
</p>

Tick the checkbox next to the provider you want to configure and insert the username and the token to use for that provider.

!!! notice "Tokens are personal"
    The tokens are personal to each user. When you perform operations on a repository, your token is going to be used, even though the repository might belong to a different user.  

## Repositories 

> Repositories are cloned and managed within the scope of a project. The content of the repository will reside on Hopsworks File System. The content of the repository can be edited from Jupyter notebooks and can be used to configure Jobs.

Reviewer: ```suggestion
Repositories are cloned and managed within the scope of a project. The content of the repository will reside on the Hopsworks File System. The content of the repository can be edited from Jupyter notebooks and can for example be used to configure Jobs.
```

Repositories can be managed from the Git section in the project settings. The Git overview in the project settings provides a list of repositories currently cloned within the project, the location of their content as well which branch and commit their HEAD is currently at. 

The list of past and ongoing activities on the repositories is also displayed. The list contains the repository on which the activity is performed, the user performing the operation, its status as well as a message describing the outcome.

<p align="center">
  <figure>
    <a  href="../../assets/images/git/repository_overview.png">
      <img src="../../assets/images/git/repository_overview.png" alt="Repository overview">
    </a>
    <figcaption>Git provider configuration</figcaption>
  </figure>
</p>

### Clone a repository

> To clone a new repository, click on the `Clone Repository` button on the Git overview page.

Reviewer: ```suggestion
To clone a new repository, click on the `Clone repository` button on the Git overview page.
```

<p align="center">
  <figure>
    <a  href="../../assets/images/git/git_clone.png">
      <img src="../../assets/images/git/git_clone.png" alt="Clone a repository">
    </a>
    <figcaption>Git clone</figcaption>
  </figure>
</p>

> The clone page asks you to specify the URL of the repository you want to clone. As mentioned above, the supported protocol is HTTPs. As an example, if the repository is hosted on Github, the URL should look like: `https://github.com/logicalclocks/hopsworks.git`.

Reviewer: ```suggestion
The clone page asks you to specify the URL of the repository you want to clone. As mentioned above, the supported protocol is HTTPS. As an example, if the repository is hosted on Github, the URL should look like: `https://github.com/logicalclocks/hopsworks.git`.
```

Additionally the UI asks you to specify which branch you want to clone. By default the UI is going to clone the `main` branch, however a different branch or commit can be specified by selecting _clone from a specific branch_.

You can select the folder, within your project, on which the repository should be cloned. By default, the repository is going to be cloned within the `Resources` dataset. However, by clicking on the location button, a different location can be selected.

> Finally, click on the `Clone Repository` button to trigger the cloning of the repository.

Reviewer: ```suggestion
Finally, click on the `Clone repository` button to trigger the cloning of the repository.
```
